boards: add support for HiFive Unleashed

katsuster opened the pull request:
This patch adds new support for SiFive HiFive Unleashed which has
SiFive FU540 SoC, DDR and some peripherals.

First version is minimum only using UART, SPI and DDR memory area.

katsuster committed:
soc: riscv: add initial support for SiFive Freedom U540

This patch adds support for SiFive Freedom U540 SoC.
First version is minimum only using UART, SPI and DDR memory area.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster committed:
boards: add support for HiFive Unleashed

This patch adds new support for SiFive HiFive Unleashed which has
SiFive FU540 SoC, DDR and some peripherals.

This is first version so not support all features of the board.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster committed:
drivers: serial: sifive: support SiFive Freedom series SoCs

This patch changes condition of 'depends on' of sifive UART driver
to support other SoCs of SiFive Freedom series.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster committed:
samples: drivers: spi_flash: fix printf warning on 64bit environment

This patch fixes printf format specifier "%u" -> "%zu" for printing
an argument of size_t on 64bit environment.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

galak reviewed (commented):
Is the Unleashed supported in qemu?

katsuster commented:
Yes, QEMU supported a part of Unleashed hardware and memory map.
But not supported all (Ex. SPI is not exist on QEMU).

```
qemu-system-riscv64 -machine sifive_u,dumpdtb=sifive_u.dtb
dtc sifive_u.dtb -I dtb -O dts -o sifive_u.dts
```

[sifive_u.dts.txt](https://github.com/zephyrproject-rtos/zephyr/files/6632297/sifive_u.dts.txt)

carlescufi commented:
@katsuster I noticed that this board is discontinued. Is there another board in production with the same SoC?

katsuster commented:
@carlescufi Yes, already discontinued. And unfortunately I don't know another boards has FU540.
SiFive is moving next generation SoC FU740 (HiFive Unmatched).

If this support is not valuable for Zephyr, please drop it.

katsuster commented:
@galak Sorry, I got a mistake. QEMU emulates SPI and GPIO of Unleashed. And can boot from Flash.
https://qemu.readthedocs.io/en/latest/system/riscv/sifive_u.html

katsuster commented:
@galak @carlescufi 
Sorry for ask again... If Zephyr project policy does not welcome the patch for a discontinued board as Unleashed, I will close this PR.
But if not, I want to progress this PR and continue to add supported device (or add driver if we need).

cfriedt commented:
@katsuster - I'm not sure what zephyr policy is w.r.t. unsupported boards, but do you have access to an Unmatched? I'm sure a PR for that board would be quite welcome and likely some of the same code could be recycled.

cwshu commented:
@cfriedt @katsuster 
According to their dtsi in linux kernel, I think most peripherals in Unleashed/Unmatched are highly compatible.
Use uart driver as example: 
- Unmatched compatible string is  "sifive,fu740-c000-uart", "sifive,uart0"
- Unleashed compatible string is "sifive,fu540-c000-uart", "sifive,uart0"

And spi/i2c/gpio/pwm/eth are the same as uart.

katsuster commented:
@cfriedt
Yes, I also have HiFive Unmatched and want to send PR for Unmatched in next PR. So if this PR dropped, I copy and modify codes for next PR. But if accepted, I just add few codes and DTS for supporting Unmatched.
@cwshu 
Thanks a lot for Linux information. I think we can reuse almost same driver codes for Unleashed and Unmatched :)

katsuster commented:
I found a issue of current OpenOCD that does not support large size SPI Flash so I'm trying to fix and upstream it.
Currently we cannot burn the Zephyr flash ROM image into SPI Flash of Unleashed and Unmatched.

After that I'll try to create patch for HiFive Unmatched.

cfriedt commented:
@katsuster - Could you add support for both in this PR? I think that would sway anyone who might be hesitant to accept

cfriedt reviewed (approved):
I'm approving on the condition that the Unmatched can be added in this PR :-)

fkokosinski commented:
Hi! I managed to get your code working in emulation on Renode (see attached [patch](https://github.com/zephyrproject-rtos/zephyr/files/6968386/0001-boards-run-HiFive-Unleashed-on-Renode.patch.txt)). I had to rebase your branch onto upstream/main, though. Maybe it would be beneficial to run this board in simulation on Renode, as it supports Ethernet/QSPI/I²C/PWM for this board?

katsuster commented on dts/riscv/riscv64-fu540.dtsi:
```diff
@@ -0,0 +1,148 @@
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+	compatible = "SiFive,FU540-C000", "fu540-dev", "sifive-dev";
```
Thanks! I will send other PR to fix FE310's same mistake.

katsuster replied:
Done.

katsuster committed:
soc: riscv: add initial support for SiFive Freedom U740

This patch adds support for SiFive Freedom U740 SoC.
First version is minimum only using UART, SPI and L2-LIM area.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster committed:
boards: add support for HiFive Unmatched

This patch adds new support for SiFive HiFive Unmatched which has
SiFive FU740 SoC, DDR and some peripherals.

This is first version so not support all features on the board.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster commented:
Add support for HiFive Unmatched.
And I will check build error of SPI flash sample.

katsuster committed:
samples: jesd216: Exclude hifive_unleashed and unmatched boards

The HiFive Unleashed and HiFive Unmatched boards do not enable the
SPI0 controller that the flash is connected to. This test will not
build on these platforms because SPI0 bus controller is missing.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster commented:
@galak @carlescufi @cfriedt
I added HiFive Unmatched, it's currently available board.
Unleashed is discontinued but as @fkokosinski said this is useful on Renode environment (and also on QEMU?).

cfriedt commented:
I would love to merge both / all 😉

fkokosinski commented on boards/riscv/hifive_unleashed/hifive_unleashed_defconfig:
```diff
@@ -0,0 +1,16 @@
+CONFIG_BOARD_HIFIVE_UNLEASHED=y
+CONFIG_RISCV_MACHINE_TIMER=y
+CONFIG_PLIC=y
+CONFIG_PINMUX=y
```
I'm getting the following warning when building either `hifive_unleashed` or `hifive_unmatched`:
```
CMake Warning at ../../CMakeLists.txt:718 (message):
  No SOURCES given to Zephyr library: drivers__pinmux

  Excluding target from build.
```

Is this intentional?

katsuster replied:
Oh, not intentional. I'll fix it. Thanks!

katsuster replied:
Fixed.

fkokosinski reviewed (approved):
LGTM
